generator/src/test/java/com/scottlogic/deg/generator/fieldspecs/FieldSpecTests.java

```diff
@@ -0,0 +1,614 @@
+        FieldSpec augmentedFieldSpec = original.withNullRestrictions(restrictions, FieldSpecSource.Empty);
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getNullRestrictions(), restrictions);
+    }
+
+    @Test
+    public void shouldCreateNewInstanceWithDateTimeRestrictions(){
+        FieldSpec original = FieldSpec.Empty;
+        DateTimeRestrictions restrictions = mock(DateTimeRestrictions.class);
+        FieldSpec augmentedFieldSpec = original.withDateTimeRestrictions(restrictions, FieldSpecSource.Empty);
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getDateTimeRestrictions(), restrictions);
+    }
+
+    @Test
+    public void shouldCreateNewInstanceWithGranularityRestrictions(){
+        FieldSpec original = FieldSpec.Empty;
+        GranularityRestrictions restrictions = mock(GranularityRestrictions.class);
+        FieldSpec augmentedFieldSpec = original.withGranularityRestrictions(restrictions, FieldSpecSource.Empty);
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getGranularityRestrictions(), restrictions);
+    }
+
+    @Test
+    public void shouldCreateNewInstanceWithFormatRestrictions(){
+        FieldSpec original = FieldSpec.Empty;
+        FormatRestrictions restrictions = mock(FormatRestrictions.class);
+        FieldSpec augmentedFieldSpec = original.withFormatRestrictions(restrictions, FieldSpecSource.Empty);
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getFormatRestrictions(), restrictions);
+    }
+
+    @Test
+    public void shouldCreateNewInstanceWithMustContainsRestrictions(){
+        FieldSpec original = FieldSpec.Empty;
+        MustContainRestriction restrictions = mock(MustContainRestriction.class);
+        FieldSpec augmentedFieldSpec = original.withMustContainRestriction(restrictions);
+
+        Assert.assertNotSame(original, augmentedFieldSpec);
+        Assert.assertSame(augmentedFieldSpec.getMustContainRestriction(), restrictions);
+    }
+
+    @Test
+    public void emptyFieldSpecsShouldBeEqualAndHaveSameHashCode(){
+        FieldSpec first = FieldSpec.Empty;
+        FieldSpec second = FieldSpec.Empty;
+
+        Assert.assertThat(first, equalTo(second));
+        Assert.assertThat(first.hashCode(), equalTo(second.hashCode()));
+    }
+
+    @Test
+    public void emptyFieldSpecAugmentedWithNothingShouldBeEqualToEmpty(){
+        FieldSpec first = FieldSpec.Empty;
+        FieldSpec augmented = FieldSpec.Empty.withNullRestrictions(null, FieldSpecSource.Empty);
+
+        Assert.assertThat(first, equalTo(augmented));
+        Assert.assertThat(first.hashCode(), equalTo(augmented.hashCode()));
+    }
+
+    @Test
+    public void fieldSpecsWithEqualSetRestrictionsShouldBeEqual(){
+        SetRestrictions aRestrictions = new MockSetRestrictions(true);
```

Comment:
checking for equality can be done with 
    
        SetRestrictions mockSetRestrictions =  mock(SetRestrictions.class);

Reply:
Unfortunately not, mockito doesn't support asserting/verifying `equals` or `hashCode`
https://github.com/mockito/mockito/wiki/FAQ